packages/admin/cms-admin/src/dam/DataGrid/FolderDataGrid.tsx

```diff
@@ -105,6 +125,91 @@ const FolderDataGrid = ({
+                id = firstFolder.id;
+
+                if (firstFolder.parentId === currentFolderId) {
+                    // upload to current folder / creates new folders
+                    parentId = currentFolderId;
+                    redirectToSubfolder = false;
+                } else {
+                    // upload to subfolder / creates new folders
+                    parentId = firstFolder.parentId;
+                    redirectToSubfolder = true;
+                }
+            } else {
+                const files = fileUploadApi.newlyUploadedItemIds;
+                const firstFile = files[0];
+
+                type = "File";
+                id = firstFile.id;
+
+                if (firstFile.parentId === currentFolderId) {
+                    // upload to current folder / creates NO new folders (only files)
+                    parentId = currentFolderId;
+                    redirectToSubfolder = false;
+                } else {
+                    // upload to subfolder / creates NO new folders (only files)
+                    parentId = firstFile.parentId;
+                    redirectToSubfolder = true;
+                }
+            }
+
+            if (id === redirectedToId) {
+                // otherwise it's not possible to navigate to another folder while the new item is in newlyUploadedItemIds
+                // because it always automatically redirects to the new item
+                return;
+            }
+
+            const result = await apolloClient.query<GQLDamItemListPositionQuery, GQLDamItemListPositionQueryVariables>({
+                query: damItemListPosition,
+                variables: {
+                    id: id,
+                    type: type,
+                    folderId: parentId,
+                    includeArchived: filterApi.current.archived,
+                    filter: {
+                        mimetypes: props.allowedMimetypes,
+                        searchText: filterApi.current.searchText,
+                    },
+                    sortColumnName: filterApi.current.sort?.columnName,
+                    sortDirection: filterApi.current.sort?.direction,
+                },
+            });
+
+            const position = result.data.damItemListPosition;
+            const targetPage = Math.floor(position / dataGridProps.pageSize);
+
+            if (redirectToSubfolder && id !== redirectedToId && parentId && parentId !== currentFolderId) {
+                switchApi.activatePage("folder", parentId);
+            } else {
+                dataGridProps.onPageChange?.(targetPage, {});
+            }
+
+            setRedirectedToId(id);
+        }
+
+        navigateToNewlyUploadedItems();
+
+        // useEffect dependencies must only include `newlyUploadedItemIds`, because the function should only be called once after new items are added.
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [fileUploadApi.newlyUploadedItemIds]);
```

Comment:
The correct way would be to keep an copy of the  uploaded items and compare them. But it's possible okay this way too